### PR TITLE
[data][doc] Document 2 datasets in 1 cluster usage

### PR DESCRIPTION
**Pull request by TimothySeah**
# Summary

Document how to run 2 datasets in 1 cluster for various use cases including training + validation.

We actually need to set this in two places for async validation:
1) We recommend creating the dataset in a copy of the current data context that contains the label selector. This scopes creation-time tasks like `read_parquet` to the cluster
2) (only applies to ray train; ray data doesn't need this) we also need to set the `DataConfig` because Ray Train overrides the Dataset's options at training start. Maybe we can make this more ergonomic in a future PR.

### CODE REVIEW

File: doc/source/train/doc_code/asynchronous_validation.py [hunk 1]
```diff
@@ -144,12 +168,35 @@ def train_func(config: dict) -> None:
...
+    # Set the global ``DataContext`` to the "training" subcluster BEFORE
+    # creating the training Dataset so reads/schema-inference tasks land
+    # on training nodes. ``Dataset.context`` is a deep copy of the global
+    # taken at construction, so this single setting also propagates to
+    # every operator the Dataset later launches. See
+    # https://docs.ray.io/en/latest/data/concurrent-dataset-execution.html.
+    ray.data.DataContext.get_current().execution_options.label_selector = {
+        "ray-subcluster": "training"
+    }
     train_dataset = ray.data.read_parquet(...)
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply doc/source/train/doc_code/asynchronous_validation.py]: I think this is better^

File: doc/source/data/concurrent-dataset-execution.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import ray
+
+# Set the global DataContext first.
+ctx = ray.data.DataContext.get_current()
+ctx.execution_options.label_selector = {"ray-subcluster": "tenant_a"}
+
+# Now create the Dataset. ``Dataset.context`` is a deep copy of the global
+# ``DataContext`` taken at construction time, so it inherits the selector
+# automatically — you don't need to also set ``dataset.context``.
+dataset = ray.data.read_parquet("s3://my-bucket/tenant_a/")
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply doc/source/data/concurrent-dataset-execution.md]: done.

File: doc/source/data/concurrent-dataset-execution.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ctx = ray.data.DataContext.get_current()
+
+# Tenant A: subcluster "tenant_a".
+ctx.execution_options.label_selector = {"ray-subcluster": "tenant_a"}
+ds_a = ray.data.read_parquet("s3://my-bucket/tenant_a/")
+
+# Tenant B: subcluster "tenant_b". Set the global context again
+# right before creating ds_b.
+ctx.execution_options.label_selector = {"ray-subcluster": "tenant_b"}
+ds_b = ray.data.read_parquet("s3://my-bucket/tenant_b/")
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply doc/source/data/concurrent-dataset-execution.md]: done.

File: doc/source/train/doc_code/asynchronous_validation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TimothySeah) [reply doc/source/train/doc_code/asynchronous_validation.py]: Agreed - this is basically the contextmanager approach we discussed berfore.

**Review by justinvyu — APPROVED** [review]
🥇

**Review by elliot-barn — APPROVED** [review]
